cirq-core/cirq/devices/grid_qubit.py

```diff
@@ -178,22 +171,33 @@ class GridQid(_BaseGridQid):
     cirq.GridQid(5, 4, dimension=2)
     """
 
-    def __init__(self, row: int, col: int, *, dimension: int) -> None:
-        """Initializes a grid qid at the given row, col coordinate
+    _cache = weakref.WeakValueDictionary[Tuple[int, int, int], 'cirq.GridQid']()
+
+    def __new__(cls, row: int, col: int, *, dimension: int) -> 'cirq.GridQid':
+        """Creates a grid qid at the given row, col coordinate
 
         Args:
             row: the row coordinate
             col: the column coordinate
             dimension: The dimension of the qid's Hilbert space, i.e.
                 the number of quantum levels.
         """
-        self.validate_dimension(dimension)
-        self._row = row
-        self._col = col
-        self._dimension = dimension
+        key = (row, col, dimension)
+        inst = cls._cache.get(key)
+        if inst is None:
+            cls.validate_dimension(dimension)
+            inst = super().__new__(cls)
+            inst._row = row
+            inst._col = col
+            inst._dimension = dimension
+            cls._cache[key] = inst
+        return inst
+
+    def __getnewargs_ex__(self):
```

Comment:
This seems like a pretty obscure dunder function.  Maybe a docstring to explain what this is would be helpful.